lib/core/src/Cardano/Wallet/DB/Sqlite.hs

```diff
@@ -258,28 +260,33 @@ newDBFactory tr defaultFieldValues mDatabaseDir = do
                 traceWith tr $ MsgRemoving (pretty wid)
                 modifyMVar_ mvar (pure . Map.delete wid)
             }
-        Just databaseDir -> pure DBFactory
-            { withDatabase = \wid action -> do
-                withDBLayer
-                    tr
-                    defaultFieldValues
-                    (Just $ databaseFile wid)
-                    (action . snd)
+
+    Just databaseDir -> do
+        refs <- newRefCount
+        pure DBFactory
+            { withDatabase = \wid action -> withRef refs wid $ withDBLayer
+                tr
+                defaultFieldValues
+                (Just $ databaseFile wid)
+                (action . snd)
             , removeDatabase = \wid -> do
-                traceWith tr $ MsgRemoving (pretty wid)
-                let files =
-                        [ databaseFile wid
-                        , databaseFile wid <> "-wal"
-                        , databaseFile wid <> "-shm"
-                        ]
-                mapM_ removePathForcibly files
+                let widp = pretty wid
+                -- try to wait for all 'withDatabase' calls to finish before
+                -- deleting database file.
+                let trWait = contramap (MsgWaitingForDatabase widp) tr
+                waitForFree trWait refs wid $ \inUse -> do
```

Comment:
Why do we need the reference counting now, but didn't previously?

Only using `deleteSqliteDatabase`, without `waitForFee`, wouldn't achieve the same thing? (which is, I guess, to make sure the file is really gone when `removeDatabase` returns)

Some more wandering questions from trying to understand:

Is `withDatabase` is called once by the restoration worker? Then I guess `waitForFree` makes sure that the wallet is not in the registry. _And_ it looks like `destroyDBLayer` makes sure to close the sqlite connection before it is removed from the registry.

So we could never cancel a sqlite query, but theoretically an API request could acquire a `SqliteContext` when it is open, wait, and when it tries to use it it has been closed?

Reply:
> Why do we need the reference counting now, but didn't previously?

We actually always needed it. It has never worked properly on Windows :)